File: java/src/main/java/com/myriad/wdl/model/statements/WdlConditional.java
package com.myriad.wdl.model.statements;

import com.myriad.wdl.model.base.WdlNode;
import com.myriad.wdl.model.definitions.WdlWorkflow.WdlWorkflowElement;
import com.myriad.wdl.model.expressions.WdlExpression;
import java.util.ArrayDeque;
import lombok.Getter;
import lombok.Setter;

/**
 * Conditional statement node.
 *
 * <p>This models `if`, `else if`, and `else` workflow control flow exactly as written in the
 * source. See the WDL specification section "Conditional Statement".
 */
public class WdlConditional implements WdlStatement, WdlWorkflowElement {

  @Getter @Setter private WdlExpression condition;
  private final ArrayDeque<WdlStatement> thenStatements = new ArrayDeque<>();
  private final ArrayDeque<WdlConditionalElseIf> elseIfs = new ArrayDeque<>();
  private final ArrayDeque<WdlStatement> elseStatements = new ArrayDeque<>();

  public WdlConditional() {}

  /** Returns the ordered statements in the `then` branch. */
  public ArrayDeque<WdlStatement> thenStatements() {
    return thenStatements;
  }

  /** Returns the ordered `else if` branches. */
  public ArrayDeque<WdlConditionalElseIf> elseIfs() {
    return elseIfs;
  }

  /** Returns the ordered statements in the `else` branch. */
  public ArrayDeque<WdlStatement> elseStatements() {
    return elseStatements;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.CONDITIONAL;
  }

  /** Single `else if` branch inside a conditional statement. */
  public static class WdlConditionalElseIf implements WdlNode {
    @Getter @Setter private WdlExpression condition;
    private final ArrayDeque<WdlStatement> thenStatements = new ArrayDeque<>();

    /** Returns the ordered statements in this `else if` branch. */
    public ArrayDeque<WdlStatement> thenStatements() {
      return thenStatements;
    }
  }
}

File: python/src/wdl_model/model/statements/wdl_conditional.py
"""Conditional statement nodes for the Python WDL object model."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from wdl_model.model.base import WdlNode, WdlSourceRange
from wdl_model.model.definitions import WdlWorkflowElement
from wdl_model.model.expressions import WdlExpression

from .wdl_statement import ComponentType, WdlStatement


@dataclass
class WdlConditionalElseIf(WdlNode):
    """Single ``else if`` branch inside a conditional statement."""

    condition: WdlExpression | None = None
    _thenStatements: deque[WdlStatement] = field(default_factory=deque)

    def thenStatements(self) -> deque[WdlStatement]:
        """Return the ordered statements in this ``else if`` branch."""
        return self._thenStatements


@dataclass
class WdlConditional(WdlStatement, WdlWorkflowElement):
    """Conditional statement node with ``if``, ``else if``, and ``else`` branches."""

    condition: WdlExpression | None = None
    source_range: WdlSourceRange | None = None
    _thenStatements: deque[WdlStatement] = field(default_factory=deque)
    _elseIfs: deque[WdlConditionalElseIf] = field(default_factory=deque)
    _elseStatements: deque[WdlStatement] = field(default_factory=deque)

    def thenStatements(self) -> deque[WdlStatement]:
        """Return the ordered statements in the ``then`` branch."""
        return self._thenStatements

    def elseIfs(self) -> deque[WdlConditionalElseIf]:
        """Return the ordered ``else if`` branches."""
        return self._elseIfs

    def elseStatements(self) -> deque[WdlStatement]:
        """Return the ordered statements in the ``else`` branch."""
        return self._elseStatements

    def componentType(self) -> ComponentType:
        """Return the statement category for traversal and validation dispatch."""
        return ComponentType.CONDITIONAL
